hamilton/plugins/lightgbm_extensions.py

```diff
@@ -0,0 +1,100 @@
+import dataclasses
+from pathlib import Path
+from typing import Any, Collection, Dict, Optional, Tuple, Type, Union
+
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal
+
+try:
+    import lightgbm
+except ImportError:
+    raise NotImplementedError("LightGBM is not installed.")
+
+try:
+    from sklearn.exceptions import NotFittedError
+except ImportError:
+    raise NotImplementedError("scikit-learn is not installed.")
+
+
+from hamilton import registry
+from hamilton.io import utils
+from hamilton.io.data_adapters import DataLoader, DataSaver
+
+LIGHTGBM_MODEL_TYPES = [lightgbm.LGBMModel, lightgbm.Booster, lightgbm.CVBooster]
+LIGHTGBM_MODEL_TYPES_ANNOTATION = Union[lightgbm.LGBMModel, lightgbm.Booster, lightgbm.CVBooster]
+
+
+@dataclasses.dataclass
+class LightGBMFileWriter(DataSaver):
+    """Write LighGBM models and boosters to a file"""
+
+    path: Union[str, Path]
+    num_iteration: Optional[int] = None
+    start_iteration: int = 0
+    importance_type: Literal["split", "gain"] = "split"
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return LIGHTGBM_MODEL_TYPES
+
+    def save_data(self, data: LIGHTGBM_MODEL_TYPES_ANNOTATION) -> Dict[str, Any]:
+        if isinstance(data, lightgbm.LGBMModel):
+            try:
+                data.booster_.save_model(
+                    filename=self.path,
+                    num_iteration=self.num_iteration,
+                    start_iteration=self.start_iteration,
+                    importance_type=self.importance_type,
+                )
+            except NotFittedError as e:
+                return utils.get_exception_details(e)
+        else:
+            data.save_model(
+                filename=self.path,
+                num_iteration=self.num_iteration,
+                start_iteration=self.start_iteration,
+                importance_type=self.importance_type,
+            )
```

Comment:
not knowing the API, would it make more sense to check if `data` has an attribute `booster_`  versus the class? Also it seems counterintuitive that you're not checking for the `Booster` model.

Otherwise it seems like we should fail if it's not fitted, rather than raising an error? what's your rationale there?

Reply:
Regarding failure, it's a question I raised in the [XGBoost PR](https://github.com/DAGWorks-Inc/hamilton/pull/459#issue-1942443037) (see initial message). Currently, the committed XGBoost extension works like this LightGBM one and propagates the package `NotFittedError` through an exception message (instead of returning metadata). It is difficult to anticipate all usage patterns, but the most likely scenario for triggering `NotFittedError` would be an incorrect dependency specified in the materializer definition (e.g., trying to materialize `base_model` instead of `trained_model`). 

IMO, the ideal approach is unclear. On one hand, if you are writing code to create and store models you would want it to fail. On the other hand, if you wrote a DAG that executes properly, adding a materializer shouldn't break your code. If materialization nodes are always executed at the end, the DAG should be able to complete before the materializer fails.

Reply:
will that NotFittedError be raised on access to the `booster_` attribute? Otherwise looks good.